deepchem/utils/poly_wd_graph_utils.py

```diff
@@ -0,0 +1,153 @@
+    This class provides methods to validate the format of a datapoint string.
+    This is a specific string format that is used for storing weighted directed polymer data in a parsable format.
+
+    The format is as follows:
+        [monomer1].[monomer2]|[fraction_of_monomer1]|[fraction_of_monomer2]|<[polymer_rule1]<[polymer_rule2]
+    The polymer rule has an own format in it. Which is as follows:
+        [[atom_index1]-[atom_index2]]:[fraction_of_bond_between_atom1_to_atom2]:[fraction_of_bond_between_atom2_to_atom1]
+
+    This format is explicitly used for formatting the input for Weighted Directed Message Passing Neural Networks (wD-MPNN).
+    The input format holds a SMART notation and regular expression formatting to keep molecular data with corresponding bonds and weights.
+    Irrespective of this explicit usecase, the formatting can allow featurization of same data for other graph based neural networks.
+
+    The validate method validates the proper formatting for monomer molecules, proper value of the fractions and valid atom indicies and corresponding weights
+    in the polymer rules.
+
+    Example
+    -------
+    >>> from deepchem.utils import PolyWDGStringValidator
+    >>> validator = PolyWDGStringValidator()
+    >>> validator.validate("[1*]C.C[2*]|0.5|0.5|<1-2:0.5:0.5")
+    True
+    """
+
+    @staticmethod
+    def get_parsed_vals(datapoint: str) -> Tuple[str, list, str]:
+        """
+        This static method parses the datapoint string into 3 parts:
+        1. Monomer molecules
+        2. Fragments
+        3. Polymer rules
+
+        Parameters
+        ----------
+        datapoint : str
+            The datapoint string to parse
+        Returns
+        -------
+        Tuple[str, list, str]
+            A tuple containing the 3 parts of the datapoint string
+        """
+        base_parsed = datapoint.split("|")
+        if len(base_parsed) < 3:
+            raise ValueError(
+                f"Invalid datapoint format: At least 3 splits should be there but found {len(base_parsed)} no. of splits"
+            )
+        monomer_mols = base_parsed[0]
+        polymer_rules = base_parsed[-1]
+        fragments = base_parsed[1:-1]
+        return monomer_mols, fragments, polymer_rules
+
+    @staticmethod
+    def get_polymer_rules(rules_str: str) -> List[str]:
+        """
+        This static method parses the polymer rules string into a list of rules.
+
+        Parameters
+        ----------
+        rules_str : str
+            The polymer rules string to parse
+
+        Returns
+        -------
+        List[str]
+            A list containing the parsed rule strings
+        """
+        if len(rules_str.split("<")) == 1:
+            raise ValueError(
+                "Invalid rules string: The rule string must contain '<' as a separator for rules !"
+            )
+        return rules_str.split("<")[1:]
+
+    def _validate_fragments(self, datapoint: str):
```

Comment:
pls add doc string in this function.

Reply:
These are private methods, hence no doc strings were added. But if it's required, I will add it.